src/libutils/oglapphelpers/msl.mm

```
@@ -0,0 +1,504 @@
+            }
+            break;
+                
+            case UNIFORM_VECTOR_FLOAT:
+            {
+                int size = data.m_vectorFloat.m_getSize();
+                const float dummyFloat = 123456789.0f;
+                const float* v = size == 0 ? &dummyFloat : data.m_vectorFloat.m_getVector();
+                size = size == 0 ? sizeof(float) : size * sizeof(float);
+                [renderCmdEncoder setFragmentBytes:v length:size atIndex:uniformId++];
+            }
+            break;
+                
+            case UNIFORM_UNKNOWN:
+                throw Exception("Unknown uniform type.");
+        };
+    }
+}
+
+bool MetalBuilder::buildPipelineStateObject(const std::string & clientShaderProgram)
+{
+    NSString* shaderSrc = [NSString stringWithUTF8String:clientShaderProgram.c_str()];
+    
+    NSError* error = nil;
+    MTLCompileOptions* options = [MTLCompileOptions new];
+    [options setLanguageVersion:MTLLanguageVersion2_0];
+    [options setFastMathEnabled:NO];
+    m_library = [[m_device newLibraryWithSource:shaderSrc options:options error:&error] autorelease];
+    
+    id<MTLFunction> vertexShader = [[m_library newFunctionWithName:@"ColorCorrectionVS"] autorelease];
+    id<MTLFunction> pixelShader  = [[m_library newFunctionWithName:@"ColorCorrectionPS"] autorelease];
+    
+    MTLRenderPipelineDescriptor* renderPipelineDesc = [[MTLRenderPipelineDescriptor new] autorelease];
+    [renderPipelineDesc setVertexFunction:vertexShader];
+    [renderPipelineDesc setFragmentFunction:pixelShader];
+    [renderPipelineDesc setVertexDescriptor:nil];
+    [renderPipelineDesc setRasterizationEnabled:YES];
+    [renderPipelineDesc setInputPrimitiveTopology:MTLPrimitiveTopologyClassTriangle];
+    [[renderPipelineDesc colorAttachments][0] setPixelFormat:MTLPixelFormatRGBA32Float];
+    m_PSO = [[m_device newRenderPipelineStateWithDescriptor:renderPipelineDesc error:&error] autorelease];
+    
+    return (error == nil);
```

The code should throw an exception to better trap the issue; otherwise, `MetalApp::redisplay()` will still try to apply a color correction even after the failure.